lib/ostruct.rb

```diff
@@ -299,11 +301,16 @@ def dig(name, *names)
   #   person.pension = nil
   #   person                 # => #<OpenStruct name="John", pension=nil>
   #
+  #   person.delete_field('number')  # => NameError
+  #
+  #   person.delete_field('number') { 8675_309 } # => 8675309
+  #
   def delete_field(name)
     sym = name.to_sym
     begin
       singleton_class.remove_method(sym, "#{sym}=")
     rescue NameError
+      return yield if block_given?
```

Comment:
I'd move that line two lines below, before the `raise`, as it seems to be the most appropriate place.
It shouldn't make any practical difference, since the existence of the setter method and the entry in the hash should be in sync, but it seems more appropriate.

Reply:
I figured there was no reason to access the hash if it was already known the desired field didn't exist. I prefer its current location, but I'll move if requested.

Reply:
I understand the reason for you placing it there. My concern is that it is not at the same place than the `raise`. It fundamentally has to be instead of the `raise`. So unless the `raise` is also moved there, yes, I strongly believe it has to move.

Reply:
Oh, actually, I think it's a bug otherwise, because OpenStruct's no longer always have a setter defined. I think `OpenStruct.new(foo: 42).delete_field(:foo) { :bug }` would return `:bug` instead of `42`, right? Sorry, no time to check it myself today

Reply:
No, it would return 42. It would only execute the provided block if an error is raised while removing from the singleton. At that point it's clear the requested field does not exist in the hash and any provided block should be executed. Tests were included to demonstrate normal functionality and expected values when a block is provided.

Reply:
Ahh, I missed the setter bit. If there is no longer always a setter then it may have the potential to give an incorrect value.